### DISCUSSION TITLE
Storage - Bucket Read Specific File vs Bucket List Files

### DISCUSSION THREAD

**Discussion by NicolasAlmerge** [supabase/supabase#26220] [Questions]
Hello,

For storage RLS, is it possible to differentiate between reading one file and listing all files within a bucket. For example, I want users to read any file as long as they have the path to that file, but not list the content of the bucket. Right now, adding a policy for select will enable both. Is it possible to only enable accessing any file given its path but not listing?

Thanks!

**Comment by GaryAustin1** [comment]
I can't think of a way.  Unlike normal tables you can't use rpc to bypass rls and get a single row.

You could make sure there is a random UUID in the pathname.  Use the public URL for that with a public bucket, and then block select so list does not work...

**Reply by GaryAustin1** [reply]
As I said, I don't know of a way.  You can block list on a public bucket with select policy, but then if someone knows or can guess a file they can get to any file one at a time.   

Maybe someone else will suggest something.  List is like select, it obeys the same RLS select policy like createSignedUrl does.
The same issue exists on tables where it is often requested to just read if you know the exact id.  There though you can use an rpc function to get that just the one id.

**Reply by GaryAustin1** [reply]
No policy means no access.  This means only public URL will be able to get the file.

Thinking about it though, you will not be able to have a user delete the file on their own because delete requires select.  So you would have to delete with service_role.  Uploads don't use select policy.   Basically you would have insert and get file with Public URL.  That might be too limiting.

I don't ever recall seeing this question.  I did glance through the issues in storage and discussions here and don't see this as a request in the past.

**Reply by GaryAustin1** [reply]
And to clarify, list would only allow users to see their own files if you use the owner column with auth.uid() as the select policy.

**Reply by GaryAustin1** [reply]
If you want to make a suggestion for the Supabase team you could suggest an enhancement here or the storage repository.   Not sure they will see this thread.

**Reply by GaryAustin1** [reply]
With the policy of select to auth.uid() then the owner will be able to list all their own files only.
